### PR TITLE
Improved Docker Registry Support 

### PR DESCRIPTION
**Pull request by ArcticSnowman**
This is an improved version of the docker registry support in PR #5 

Added Feature:
* The triggers only fire when a 'push' action is sent by docker registry not pull.

This PR rolls up the changes from PR #5

### CODE REVIEW

**Comment by ArcticSnowman** [comment]
https://issues.jenkins-ci.org/browse/JENKINS-30931

**Comment by michaelneale** [comment]
LGTM

**Comment by michaelneale** [comment]
I still wasn't able to verify this with: 

```
curl -X POST -H "Content-Type: application/json" http://localhost:8080/jenkins/dockerregistry-webhook/notify -d @src/test/resources/private-registry-payload-1-repository.json
```

**Comment by ArcticSnowman** [comment]
@michaelneale   - How are you starting jenkins?    What is the URL to the jenkins UI?

Try this if the jenkins UI is at 'http://localhost:8080/'

```curl -X POST -H "Content-Type: application/json" http://localhost:8080/dockerregistry-webhook/notify -d @src/test/resources/private-registry-payload-1-repository.json```

**Comment by michaelneale** [comment]
Using mvn hpi:run (so the extra Jenkins uri path element is needed)

**Comment by GrantStreetGroup** [comment]
Try a '-v' on the curl command line.. this is what I got.

```
* About to connect() to localhost port 18790 (#0)
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 18790 (#0)
> POST /jenkins/dockerregistry-webhook/notify HTTP/1.1
> User-Agent: curl/7.29.0
> Host: localhost:18790
> Accept: */*
> Content-Type: application/json
> Content-Length: 970
>
* upload completely sent off: 970 out of 970 bytes
< HTTP/1.1 200 OK
< Stapler-Trace-001: -> evaluate(<hudson.model.Hudson@65318dd6> :hudson.model.Hudson,"/dockerregistry-webhook/notify")
< Stapler-Trace-002: -> evaluate(((StaplerProxy)<hudson.model.Hudson@65318dd6>).getTarget(),"/dockerregistry-webhook/notify")
< Stapler-Trace-003: -> evaluate(<hudson.model.Hudson@65318dd6>.getDynamic("dockerregistry-webhook",...),"/notify")
< Stapler-Trace-004: -> evaluate(<org.jenkinsci.plugins.registry.notification.webhook.dockerregistry.DockerRegistryWebHook@f5b2be5> :org.jenkinsci.plugins.registry.notification.webhook.dockerregistry.DockerRegistryWebHook,"/notify")
< Stapler-Trace-005: -> <org.jenkinsci.plugins.registry.notification.webhook.dockerregistry.DockerRegistryWebHook@f5b2be5>.doNotify(...)
< Content-Length: 0
< Server: Jetty(6.1.1)
<
* Connection #0 to host localhost left intact
```

**Comment by michaelneale** [comment]
OK :bee:  from me. @reviewbybees @recena need this merged (this also subsumes PR #5) and released

**Comment by michaelneale** [comment]
(pull request is NOT from a cloudbees employee, that was an accident).

**Comment by rsandell** [comment]
There seems to be missing migration code for all the refactoring. Old data will fail to load.

**Comment by michaelneale** [comment]
Odd, what data should a notification plugin require?

**Comment by michaelneale** [comment]
@ArcticSnowman ^

**Comment by rsandell** [comment]
the build cause and the fingerprint data for looking up the builds when showing the landing page of the callback url for example.

**Comment by ArcticSnowman** [comment]
@michaelneale  @rsandell -  Sorry not really sure I fully understand the comment.  Is there some changes in the DockerHub side of things that are missing or not in the Docker Registry side?

**Comment by rsandell** [comment]
This plugin stores some data in various places on disk as serialized XML, just as any other Jenkins plugin does. The refactorings and renames of class files in this change will cause that data to be lost unless migration code is provided for that data.

See https://wiki.jenkins-ci.org/display/JENKINS/Hint+on+retaining+backward+compatibility

File: src/main/resources/org/jenkinsci/plugins/registry/notification/DockerHubTrigger/help.groovy [hunk 1]
```diff
@@ -21,13 +21,13 @@
...
  */
-package org.jenkinsci.plugins.dockerhub.notification.DockerHubTrigger
+package org.jenkinsci.plugins.registry.notification.DockerHubTrigger
 
 import jenkins.model.Jenkins
 
 def webHookUrl() {
     String rootUrl = Jenkins.instance?.getRootUrl() ?: "http://myJENKINS/";
-    return rootUrl + "dockerhub-webhook/notify"
+    return rootUrl + "dockerhub-webhook/nowwtify"
```
Comment (rsandell) [inline src/main/resources/org/jenkinsci/plugins/registry/notification/DockerHubTrigger/help.groovy]: It doesn't look like this has changed

**Comment by rsandell** [comment]
@ArcticSnowman @michaelneale Sorry for taking so long, I've been trying to get this on my plate for a while but now finally PR #11 is in the works.

**Comment by ArcticSnowman** [comment]
@rsandell  - Thanks,   Jenkins Plugin coding is not my are of expertise.   Should we close this PR? And then resync my fork when #11 is merged?

**Comment by rsandell** [comment]
@ArcticSnowman yes, I was planning on closing this and #5 when PR #11 is merged.

**Comment by rsandell** [comment]
It will probably be auto closed since all the commits will get to master

**Pull request merged**